src/execution/sema/sema_builtin.cpp

```diff
@@ -1038,59 +1038,142 @@ void Sema::CheckBuiltinTableIterCall(ast::CallExpr *call, ast::Builtin builtin)
+      }
+      break;
+    }
+    case ast::Builtin::TableIterIndexInsertParallel: {
+      if (!CheckArgCount(call, 7)) {
+        return;
+      }
+
+      const auto &call_args = call->Arguments();
+
+      // The first argument is a table oid.
+      if (!call_args[0]->GetType()->IsIntegerType()) {
+        ReportIncorrectCallArg(call, 0, "First argument should be an integer type.");
+        return;
+      }
+
+      // The second argument is a uint32_t array.
+      if (!call_args[1]->GetType()->IsArrayType()) {
+        ReportIncorrectCallArg(call, 1, "Second argument should be a fixed length uint32 array.");
+        return;
+      }
+      auto *arr_type = call_args[1]->GetType()->SafeAs<ast::ArrayType>();
+      if (!arr_type->GetElementType()->IsSpecificBuiltin(ast::BuiltinType::Uint32) || !arr_type->HasKnownLength()) {
+        ReportIncorrectCallArg(call, 1, "Second argument should be a fixed length uint32 array");
+      }
+
+      // The third argument is an opaque query state. For now, check it's a pointer.
+      const auto void_kind = ast::BuiltinType::Nil;
+      if (!call_args[2]->GetType()->IsPointerType()) {
+        ReportIncorrectCallArg(call, 2, GetBuiltinType(void_kind)->PointerTo());
+        return;
+      }
+
+      // The fourth argument is the execution context.
+      const auto exec_ctx_kind = ast::BuiltinType::ExecutionContext;
+      if (!IsPointerToSpecificBuiltin(call_args[3]->GetType(), exec_ctx_kind)) {
+        ReportIncorrectCallArg(call, 3, GetBuiltinType(exec_ctx_kind)->PointerTo());
+        return;
+      }
+
+      // The fifth argument is the scanner function.
+      auto *scan_fn_type = call_args[4]->GetType()->SafeAs<ast::FunctionType>();
+      if (scan_fn_type == nullptr) {
+        GetErrorReporter()->Report(call->Position(), ErrorMessages::kBadParallelScanFunction, call_args[4]->GetType());
+        return;
+      }
+      // Check the type of the scanner function parameters. See TableVectorIterator::ScanFn.
+
+      const auto tvi_kind = ast::BuiltinType::TableVectorIterator;
+      const auto index_pr_kind = ast::BuiltinType::ProjectedRow;
+      const auto inserter_kind = ast::BuiltinType::StorageInterface;
+      const auto &params = scan_fn_type->GetParams();
+      if (params.size() != 5                                              // Scan function has 5 arguments.
+          || !params[0].type_->IsPointerType()                            // QueryState, must contain execCtx.
+          || !params[1].type_->IsPointerType()                            // Thread state.
+          || !IsPointerToSpecificBuiltin(params[2].type_, tvi_kind)       // TableVectorIterator.
+          || !IsPointerToSpecificBuiltin(params[3].type_, index_pr_kind)  // TableVectorIterator.
+          || !IsPointerToSpecificBuiltin(params[4].type_, inserter_kind)  // TableVectorIterator.
+      ) {
+        GetErrorReporter()->Report(call->Position(), ErrorMessages::kBadParallelScanFunction, call_args[4]->GetType());
```

Comment:
Can you make your own error message / actually, can you move this logic to your own function instead of mixing it with parallel scan?